### PR TITLE
[GEP-20] Update the existing recovery mechanisms in the proposal

### PR DESCRIPTION
**Pull request by ialidzhikov**
**How to categorize this PR?**

/area high-availability
/kind enhancement

**What this PR does / why we need it**:
Previously there was the misunderstanding that in case of a Node/zone outage Pods on unhealthy Nodes hang forever in `Terminating` state and that we have to enhance/implement garbage collection logic for this case.
This PR updates the existing recovery mechanisms and describes the observations from https://github.com/gardener/gardener/issues/6529#issuecomment-1253517049 and https://github.com/gardener/gardener/issues/6529#issuecomment-1254800255.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/6529

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: docs/proposals/20-ha-control-planes.md [hunk 1]
```diff
@@ -528,10 +529,11 @@ In this option existing recovery mechanisms as described above are used. There i
...
-* Existing recovery mechanisms are leveraged.
+* Existing recovery mechanisms are leveraged:
+  * `kube-controller-manager` will create new replicas for the affected Deployment Pods and they will recover in short period of time.
+  * The replacement Pods for the affected StatefulSet Pods will fail to be scheduled due to their affinity rules (etcd Pods) or volume requirements (prometheus and loki Pods) to run in the outage zone and will not recover.
 
 _Cons:_
-* Existing recovery of pods will result in a downtime of up to a total of [--machine-health-timeout](https://github.com/gardener/gardener-extension-provider-gcp/blob/dff3d2417ff732fcce69ba10bbe5d04e13781539/charts/internal/machine-controller-manager/seed/templates/deployment.yaml#L49) + [--machine-drain-timeout](https://github.com/gardener/gardener-extension-provider-gcp/blob/dff3d2417ff732fcce69ba10bbe5d04e13781539/charts/internal/machine-controller-manager/seed/templates/deployment.yaml#L48) (which as of today is 2hr10min) for pods which can be rescheduled onto another node/zone. Stateful pods due to the affinity rules will not recover, however stateless pods will eventually recover.
```
Comment (ialidzhikov) [inline docs/proposals/20-ha-control-planes.md:534]: @unmarshall I decided to remove this point as it does not make sense to me. Unhealthy Nodes should not be drained, instead they should be deleted after `--machine-health-timeout` (10min by default). Hence the recovery time should not be > 2h. Let me know if I am missing something.

cc @himanshu-kun

File: docs/proposals/20-ha-control-planes.md [hunk 2]
```diff
@@ -508,10 +508,11 @@ taints:
...
+For StatefulSets, once a Pod managed by a StatefulSet transitions to `Terminating` state the `kube-controller-manager` waits until the Pod is removed from store and only after that it creates the replacement Pod. In case of a Seed node/zone outage the StatefulSet Pods in the Shoot control plane stay in `Terminating` state for 5min. After 5min the `shoot-care-controller` of `gardenlet` deletes forcefully (garbage collects) the `Terminating` Pods from the Shoot control plane. Alternatively, `machine-controller-manager` deletes the unhealthy Nodes after `--machine-health-timeout` (10min by default) and the `Terminating` Pods are removed from store shortly after the Node removal. `kube-controller-manager` creates new StatefulSet Pods. Depending on the outage type (Node or zone) the new StatefulSet Pods respectively succeed or fail to recover.
```
Comment (timuthy) [inline docs/proposals/20-ha-control-planes.md]:
```suggestion
For StatefulSets, once a Pod managed by a StatefulSet transitions to `Terminating` state the `kube-controller-manager` waits until the Pod is removed from store and only after that it creates the replacement Pod. In case of a Seed node/zone outage the StatefulSet Pods in the Shoot control plane stay in `Terminating` state for 5min. After 5min the `shoot-care-controller` of `gardenlet` deletes forcefully (garbage collects) the `Terminating` Pods from the Shoot control plane. Alternatively, `machine-controller-manager` deletes the unhealthy Nodes after `--machine-health-timeout` (10min by default) and the `Terminating` Pods are removed from store shortly after the Node removal. `kube-controller-manager` creates new StatefulSet Pods. Depending on the outage type (`node` or `zone`) the new StatefulSet Pods respectively succeed or fail to recover. For `node` the Pods will likely recover, whereas for `zone` it's usual that Pods remain in `Pending` state as long as the outage lasts, since depending volumes cannot be moved across availability zones.
```

File: docs/proposals/20-ha-control-planes.md [hunk 3]
```diff
@@ -528,10 +529,11 @@ In this option existing recovery mechanisms as described above are used. There i
 _Pros:_
 * Less complex to implement since no dynamic re-balancing of pods is required and there is no need to determine if there is an AZ outage.
 * Additional cost to host an HA shoot control plane is kept to the bare minimum.
-* Existing recovery mechanisms are leveraged.
+* Existing recovery mechanisms are leveraged:
+  * `kube-controller-manager` will create new replicas for the affected Deployment Pods and they will recover in short period of time.
+  * The replacement Pods for the affected StatefulSet Pods will fail to be scheduled due to their affinity rules (etcd Pods) or volume requirements (prometheus and loki Pods) to run in the outage zone and will not recover.
```
Comment (timuthy) [inline docs/proposals/20-ha-control-planes.md]:
```suggestion
  * The replacement Pods for the affected StatefulSet Pods will fail to be scheduled due to their affinity rules (`etcd` Pods) or volume requirements (`prometheus` and `loki` Pods) to run in the outage zone and will not recover.
```

**Review by timuthy — CHANGES_REQUESTED** [review]
Many thanks for the improvements. Just two suggestions from my side.

File: docs/proposals/20-ha-control-planes.md (hunk 1 above)
Comment (himanshu-kun) [reply docs/proposals/20-ha-control-planes.md:534]: the reasons of unhealthiness like `nodePressure` and `diskPressure` would mean that the pods would still be running on the node (except maybe a few) , so we should be doing graceful draining of node in such scenario respecting their PDBs for service availability. I am looking at it more from the user point of view when they run their workload on a cluster.
only in scenarios like `readOnlyfileSystem`(pod can't interact with fs) or `kubelet gone for too long` (eviction can't go through) are the cases where we skip drain to avoid wasting time on drain

Comment (ialidzhikov) [reply docs/proposals/20-ha-control-planes.md:534]: Thanks for explaining. In case of a zone outage I think that the Nodes will be `NotReady`/`Unknown` for a reason that is not NodePressure/DiskPressure. In the described simulations we forbid ingress/egress traffic for a zone and kubelets cannot report that they are healthy. So in this case the drain will be skipped.
Anyways, I reworked this point and removed the part with the `--machine-drain-timeout`. In today's head, it is written like it will always result in 2hr10min which should not be true.
tl;dr: If the Node is `NotReady` (except conditions like NodePressure/DiskPressure) the drain is skipped. In case of NodePressure/DiskPressure I would expect then the drain to be able to complete and not to get stuck.

**Comment by rfranzke** [comment]
@timuthy Can we merge this?

File: docs/proposals/20-ha-control-planes.md [hunk 4]
```diff
@@ -528,10 +529,12 @@ In this option existing recovery mechanisms as described above are used. There i
 _Pros:_
 * Less complex to implement since no dynamic re-balancing of pods is required and there is no need to determine if there is an AZ outage.
 * Additional cost to host an HA shoot control plane is kept to the bare minimum.
-* Existing recovery mechanisms are leveraged.
+* Existing recovery mechanisms are leveraged:
+  * `kube-controller-manager` will create new replicas for the affected Deployment Pods and they will recover in short period of time.
+  * The replacement Pods for the affected StatefulSet Pods will fail to be scheduled due to their affinity rules (`etcd` Pods) or volume requirements (`prometheus` and `loki` Pods) to run in the outage zone and will not recover.
```
Comment (timuthy) [inline docs/proposals/20-ha-control-planes.md]: I think it's quite irritating to list this a a "pro" argument, so it's rather a "con"?

File: docs/proposals/20-ha-control-planes.md [hunk 5]
```diff
@@ -528,10 +529,12 @@ In this option existing recovery mechanisms as described above are used. There i
...
 
 _Cons:_
-* Existing recovery of pods will result in a downtime of up to a total of [--machine-health-timeout](https://github.com/gardener/gardener-extension-provider-gcp/blob/dff3d2417ff732fcce69ba10bbe5d04e13781539/charts/internal/machine-controller-manager/seed/templates/deployment.yaml#L49) + [--machine-drain-timeout](https://github.com/gardener/gardener-extension-provider-gcp/blob/dff3d2417ff732fcce69ba10bbe5d04e13781539/charts/internal/machine-controller-manager/seed/templates/deployment.yaml#L48) (which as of today is 2hr10min) for pods which can be rescheduled onto another node/zone. Stateful pods due to the affinity rules will not recover, however stateless pods will eventually recover.
+* Existing recovery of pods will result in a downtime of 5min. For pods that are not part of the Shoot control plane it will be 10min (`--machine-health-timeout`).
```
Comment (timuthy) [inline docs/proposals/20-ha-control-planes.md]: Where exactly do the `5min` come from, from Gardener's garbage collection? Because then it's only the max downtime, stateless pods usually recover faster, IIUC.

**Review by timuthy — CHANGES_REQUESTED** [review]
Two more points I noticed after the changes.

File: docs/proposals/20-ha-control-planes.md (hunk 4 above)
Comment (ialidzhikov) [reply docs/proposals/20-ha-control-planes.md]: To my understanding the point here is that `Existing recovery mechanisms are leveraged` is the "pro" argument - no need to develop more sophisticated options and the existing mechanism is good enough. 

Do you suggest moving 
> * The replacement Pods for the affected StatefulSet Pods will fail to be scheduled due to their affinity rules (`etcd` Pods) or volume requirements (`prometheus` and `loki` Pods) to run in the outage zone and will not recover.

to "cons"?

I added these 2 points to clarify further what will recover (Deployment Pods) and what will not (StatefulSet Pods).

File: docs/proposals/20-ha-control-planes.md (hunk 5 above)
Comment (ialidzhikov) [reply docs/proposals/20-ha-control-planes.md]: I now added a note to clarify this.

File: docs/proposals/20-ha-control-planes.md (hunk 4 above)
Comment (timuthy) [reply docs/proposals/20-ha-control-planes.md]: The two points make the situation very transparent which is fine, but unfortunately saying "The replacement Pods for the affected StatefulSet Pods will fail to be scheduled due to their affinity rules ..." doesn't read as a "pro" argument anymore, does it? (Readers will probably think: "Why is that a positive thing?")
Maybe we can compensate this negative property by saying that `etcd` will still be operational because of its quorum and we can accept the downtime of logging/monitoring components for the time of an ongoing AZ outage.

File: docs/proposals/20-ha-control-planes.md (hunk 5 above)
Comment (timuthy) [reply docs/proposals/20-ha-control-planes.md]: Strictly, I wonder if the `5mins` are reflecting the reality.

There are still related KCM properties:
- `--node-monitor-period` (Gardener default `2min`) which makes the node turning into `Unknown`.
- `--pod-eviction-timeout` (Gardener default `2min`) which makes KCM to evict the pod after nodes remain `Unknown` for the configured duration.

Both must be passed until the pod gets a deletion timestamp which marks the `5min` start for the `shoot-care-collector`.

Comment (ialidzhikov) [reply docs/proposals/20-ha-control-planes.md]: > Both must be passed until the pod gets a deletion timestamp which marks the `5min` start for the `shoot-care-collector`.

Yes, nice catch. In general, this garbage collection in the shoot-care-controller was not added as any recovery mechanism and I am really disappointed that I have to document it in this section because it is messing with Terminating Pods and force deleting them after 5min. 😞 

In general, the whole statement did not make much sense. Above we say that Deployment Pods will recover in `short period of time` and in this sentence we try to give concrete downtime times. I reworked the points with the latest force push.

From https://github.com/gardener/gardener/pull/6732#discussion_r997022211, it should be `--node-monitor-grace-period` (2min) and not `--node-monitor-period` (5s).

Wrt to `--pod-eviction-timeout`, according to my testing and reading of KCM's source, `--pod-eviction-timeout` is only used when the taint manager is not enabled. In K8s 1.23 the taint manager is enabled by default and KCM is started with `--enable-taint-manager="true"`. When the taint manager is used, actually the time it takes to evict the affected Pods is `--node-monitor-grace-period` (2min) + 5min = 7min. I don't know yet from where these 5min come from.

For now I adapted it to be `--node-monitor-grace-period` + `--pod-eviction-timeout` but let me investigate from where these 5min come from when the taint manager is used.

Ref https://github.com/kubernetes/kubernetes/blob/v1.23.12/pkg/controller/nodelifecycle/node_lifecycle_controller.go

File: docs/proposals/20-ha-control-planes.md (hunk 4 above)
Comment (ialidzhikov) [reply docs/proposals/20-ha-control-planes.md]: I now incorporated this feedback.

File: docs/proposals/20-ha-control-planes.md (hunk 5 above)
Comment (himanshu-kun) [reply docs/proposals/20-ha-control-planes.md]: @ialidzhikov this issue discusses about the `pod-eviction-timeout` and `taint-based-eviction` thoroughly https://github.com/kubernetes/kubernetes/issues/74651
It says that `taint-based-eviction` has been enabled from `k8s 1.13` which I verified.
They have discussed ways to evict pod sooner than 5min per pod also , or there is a global flag which could be set using `kube-apiserver` flag

Comment (ialidzhikov) [reply docs/proposals/20-ha-control-planes.md]: Thank you, @himanshu-kun! It would probably take me a lot of time to find out this. 👍 

I now updated the section describing when KCM starts evicting the affected Pods from unhealthy Nodes.

**Review by timuthy — COMMENTED** [review]
Thanks a lot for all the efforts and clarification about this topic @ialidzhikov  and @himanshu-kun ❤️ 
/lgtm

**Review by timuthy — COMMENTED** [review]
/lgtm
